src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -240,6 +240,27 @@ private BoundAttribute BindAttributeCore(AttributeSyntax node, NamedTypeSymbol a
                     {
                         Error(diagnostics, ErrorCode.ERR_AttributeCtorInParameter, node, attributeConstructor.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
                     }
+
+                    for (var i = 0; i < boundConstructorArguments.Length; i++)
+                    {
+                        var parameter = GetCorrespondingParameter(i, attributeConstructor.Parameters, argsToParamsOpt, expanded);
+                        if (parameter is null)
+                        {
+                            continue;
+                        }
+
+                        var parameterType = (expanded && parameter is { IsParams: true, Type: ArrayTypeSymbol { IsSZArray: true } arrayType }) ? arrayType.ElementType : parameter.Type;
```

Comment:
>var parameterType = (expanded && parameter is { IsParams: true, Type: ArrayTypeSymbol { IsSZArray: true } arrayType }) ? arrayType.ElementType : parameter.Type; [](http://example.com/codeflow?start=24&length=160)

It doesn't look like parameter type, or whether it is `params` should really matter. For example, it looks like `object` parameter can accept an array of enum values as the value, `object[]` parameter can accept explicit array, even if it is `params`. It feels like we should simply dig through the typed constants and report an error if it uses function pointer type in any type (even in an array type, etc) on any level. And we should be able to share this check with named arguments.

Reply:
> It doesn't look like parameter type

This statement is probably too strong. The target type might matter, but at specific depth.

Reply:
Here is an example of a scenario, which is likely still broken
```
                class A : System.Attribute
                {
                    public A(object e) { }
                }

                struct B<T>
                {
                    public enum E
                    {
                        Y,
                        X
                    }
                }

                [A(new B<delegate*<void>[]>.E[]{})]
                class C { }
```

Reply:
Thanks, you are right. I added a bound tree visitor to detect funcptr-enums.

> And we should be able to share this check with named arguments.

I cannot completely share the check. The problem is that named arguments are slightly different. They serialize the type even if it's the same as type of the property/field:

https://github.com/dotnet/roslyn/blob/273955e0736451e506366ab1bfe746150ec11435/src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs#L3393-L3395

whereas constructor arguments only serialize the type when there's a conversion to `object` (or `object[]` which I missed previously):

https://github.com/dotnet/roslyn/blob/273955e0736451e506366ab1bfe746150ec11435/src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs#L3378